Doc/c-api/exceptions.rst

```diff
@@ -125,6 +125,41 @@ For convenience, some of these functions will always return a
 
    .. versionadded:: 3.5
 
+.. c:function:: void PyErr_ChainExceptions(PyObject *exc, PyObject *val, PyObject *tb)
+
+   Set the provided exception as the cause for the current exception. If there is no exception
+   currently set (:c:func:`PyErr_Occurred` returns ``0``) this function behaves as like :c:func:`PyErr_Restore`, but if an exception is already set, set the context
+   associated with it.
+
+   The exception type should be a class. Do not pass an invalid exception type or value. This call
+   takes away a reference to each object: you must own a reference to each object before the call
+   and after the call you no longer own these references.
+
+   .. warning::
+
+      The caller is responsible for ensuring that this call won't create
+      any cycles in the exception context chain.
```

Comment:
I think the doc should say what happens if you violate the rule.  I think in reality it will just create the cycle and if it doesn't bother you than good for you.

Reply:
> I think the doc should say what happens if you violate the rule.

Is undefined behavior, technically as the contract of this function withing CPython is that cycles should not be created. Some parts of the exception machinery can break (or at least they used to) in presence of cycles.

Reply:
But when the exception machinery breaks in the presence of a cycle we consider that to be a bug in python.

Reply:
How is the caller supposed to ensure this? Is there any API for checking it?

Reply:
There is no API to check this, but to create a cycle with exceptions it needs to be s deliberate action. There is not an easy way to accidentally crate an exception cycle

Reply:
> But when the exception machinery breaks in the presence of a cycle we consider that to be a bug in python.

I think the movement is in the direction of trying to prevent / avoid the creation of cycles in the first place (as was mentioned in the discussion for this [recent ticket](https://bugs.python.org/issue25782), for example). The warning being added here is in line with that.

Reply:
As per my comment below, the easiest way to violate this requirement is to call the API without setting a new exception first (or clearing the existing one). Buggy code (using my API naming suggestion from above):

```
PyErr_Fetch(&type, &value, &traceback);
// Exception state is not cleared, no new exception is set
PyErr_RestoreOrChain(type, value, traceback); // Sets exception as its own context
```

Because the exception might not have been instantiated yet when `PyErr_Fetch()` is called, there's no easy way for the API to reliably detect this error, so the onus is on the caller to manage the exception state correctly.